pkg/agent/httpmux.go

```diff
@@ -0,0 +1,117 @@
+package agent
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+
+	"go.uber.org/zap"
+)
+
+// HttpMuxServer provides a way to multiplex multiple http.ServeMux
+// instances on a single HTTP server, so that requests are routed to
+// the correct ServeMux based on a prefix at the beginning of the URL
+// path.
+//
+// For example, if you call RegisterMux("foo", myMux), then all requests to
+// "/foo/*" are routed to 'myMux' instance
+//
+// (Why is this needed? You could register all the routes directly in
+// one giant http.ServeMux, but http.ServeMux doesn't provide any way
+// to unregister patterns.)
+type HttpMuxServer struct {
+	muxedServers map[string]*http.ServeMux
+	lock         sync.Mutex // protects muxedServers
+	logger       *zap.Logger
+}
+
+// Implements http.Handler
+func (s *HttpMuxServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	s.logger.Debug(fmt.Sprintf("MUX: received request: %s", r.URL.Path))
```

Comment:
Heads up: this won't get output in production as it's below `INFO`.

Reply:
I don't know if we necessarily want to see it either, but it is probably a good idea to have logs for every communication somewhere.

Reply:
Yeah, that was on purpose. We have a decent amount of logging in the handler functions already, I didn't want to make the agent more noisy than it is now.

I added Warn-level logging to the error cases though, so that we get a trace of those in the agent log too.